### PR TITLE
[WIP]Fixed GC issue and re-factor dictionary based result collector

### PR DESCRIPTION
**Pull request by kumarvishal09**
**Problem:**When time stamp or date type is getting selected in query projection we are creating direct dictionary result collector object every for each time this is causing lots of gc.
**Solution:** Need to create only one object for time stamp and date type

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/738/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/739/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/740/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/741/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/742/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/743/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/744/

File: core/src/main/java/org/apache/carbondata/core/scan/collector/impl/DictionaryBasedResultCollector.java [hunk 1]
```diff
@@ -37,64 +37,138 @@
...
-    QueryDimension[] queryDimensions = tableBlockExecutionInfos.getQueryDimensions();
+  public DictionaryBasedResultCollector(BlockExecutionInfo blockExecutionInfos) {
+    super(blockExecutionInfos);
+    this.isMsrsPresent = measureDatatypes.length > 0;
+    this.queryDimensions = tableBlockExecutionInfos.getQueryDimensions();
     List<Integer> dictionaryIndexes = new ArrayList<Integer>();
     for (int i = 0; i < queryDimensions.length; i++) {
-      if(queryDimensions[i].getDimension().hasEncoding(Encoding.DICTIONARY) ||
-          queryDimensions[i].getDimension().hasEncoding(Encoding.DIRECT_DICTIONARY) ) {
+      if (queryDimensions[i].getDimension().hasEncoding(Encoding.DICTIONARY) || queryDimensions[i]
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/scan/collector/impl/DictionaryBasedResultCollector.java]: move queryDimensions to next line

File: core/src/main/java/org/apache/carbondata/core/scan/collector/impl/DictionaryBasedResultCollector.java [hunk 2]
```diff
@@ -129,19 +203,20 @@ public DictionaryBasedResultCollector(BlockExecutionInfo blockExecutionInfos) {
...
             }
-          } else if (directDictionaryEncodingArray[i]) {
-            DirectDictionaryGenerator directDictionaryGenerator =
-                DirectDictionaryKeyGeneratorFactory
-                    .getDirectDictionaryGenerator(queryDimensions[i].getDimension().getDataType());
-            if (directDictionaryGenerator != null) {
-              row[order[i]] = directDictionaryGenerator.getValueFromSurrogate(
+          } else if (directDictionaryEncodingArray[i] == 1
+              || directDictionaryEncodingArray[i] == 2) {
+            if (directDictionaryEncodingArray[i] == 1) {
```
Comment (jackylk) [inline core/src/main/java/org/apache/carbondata/core/scan/collector/impl/DictionaryBasedResultCollector.java]: this if check can be merged into last one

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/781/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/782/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/783/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/784/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/785/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/786/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/787/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/788/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/791/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/792/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/800/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/802/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/805/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/806/

**Pull request closed without merging**